# JSON avec Python

## Import le package JSON

In [1]:
import json

## Indiquer le chemin du fichier JSON

In [2]:
my_json_example = "example.json"

## Ouvrir un fichier JSON (et stocker son contenu dans une variable)

###Version simple

In [3]:
with open(my_json_example, 'r') as f:
    data = json.load(f)

### Version améliorée

In [4]:
def open_json_file(input_file_path):
    """Process a JSON file and call the success callback on successful processing.
    
    Args:
        input_file_path (str): Path to the input JSON file.
        output_file_path (str): Path to the output JSON file.
        success_callback (function): Function to call on successful processing.
    """

    try:
        with open(input_file_path, 'r') as f:
            data = json.load(f)
                    
        if not isinstance(data, list):
            raise ValueError("The input JSON must be an array of objects.")

    except (FileNotFoundError, json.JSONDecodeError) as e:
        print(f"Error reading the input file: {e}")
    except ValueError as e:
        print(f"Error in the data format: {e}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
    return data

In [5]:
my_data = open_json_file(my_json_example)

In [6]:
print(my_data)

[{'id': 12345, 'name': 'John Doe', 'email': 'john.doe@example.com', 'is_active': True, 'age': 30, 'addresses': [{'type': 'home', 'address': '123 Main St', 'city': 'Springfield', 'state': 'IL', 'postal_code': '62704', 'country': 'USA'}, {'type': 'work', 'address': '456 Business Rd', 'city': 'Metropolis', 'state': 'NY', 'postal_code': '10001', 'country': 'USA'}], 'phone_numbers': [{'type': 'home', 'number': '+1-555-123-4567'}, {'type': 'work', 'number': '+1-555-987-6543'}], 'preferences': {'contact_method': 'email', 'languages': ['en', 'es', 'fr'], 'timezone': 'America/Chicago'}, 'created_at': '2022-01-15T08:00:00Z', 'updated_at': '2022-06-01T12:00:00Z'}, {'id': 67890, 'name': 'Jane Smith', 'email': 'jane.smith@example.com', 'is_active': False, 'age': 25, 'addresses': [{'type': 'home', 'address': '789 Park Ave', 'city': 'New York', 'state': 'NY', 'postal_code': '10021', 'country': 'USA'}], 'phone_numbers': [{'type': 'mobile', 'number': '+1-555-234-5678'}], 'preferences': {'contact_method

## Trier les objets de premier niveau d'un fichier JSON selon une clé, par ordre alphabétique

In [7]:
#tag::sort_json_file_by_alphabetical_order[]
def sort_json_file_by_alphabetical_order(data, key):
    """Sort a JSON file by the specified key in alphabetical order and write to a new file.

    Args:
        input_file_path (str): Path to the input JSON file.
        output_file_path (str): Path to the output JSON file.
        key (str): The key to sort the JSON objects by.
    """

    sorted_data = sorted(data, key=lambda item: item[key])
    return sorted_data
#end::sort_json_file_by_alphabetical_order[]

In [8]:
sort_json_file_by_alphabetical_order(my_data, "name")

[{'id': 67890,
  'name': 'Jane Smith',
  'email': 'jane.smith@example.com',
  'is_active': False,
  'age': 25,
  'addresses': [{'type': 'home',
    'address': '789 Park Ave',
    'city': 'New York',
    'state': 'NY',
    'postal_code': '10021',
    'country': 'USA'}],
  'phone_numbers': [{'type': 'mobile', 'number': '+1-555-234-5678'}],
  'preferences': {'contact_method': 'phone',
   'languages': ['en'],
   'timezone': 'America/New_York'},
  'created_at': '2021-05-10T14:30:00Z',
  'updated_at': '2022-05-01T08:00:00Z'},
 {'id': 12345,
  'name': 'John Doe',
  'email': 'john.doe@example.com',
  'is_active': True,
  'age': 30,
  'addresses': [{'type': 'home',
    'address': '123 Main St',
    'city': 'Springfield',
    'state': 'IL',
    'postal_code': '62704',
    'country': 'USA'},
   {'type': 'work',
    'address': '456 Business Rd',
    'city': 'Metropolis',
    'state': 'NY',
    'postal_code': '10001',
    'country': 'USA'}],
  'phone_numbers': [{'type': 'home', 'number': '+1-555-12

## Trier les clés de premier niveau par ordre alphabétique

## Supprimer des doublons selon une clé

Exemple :

On a un objet A :

{
    "id": 12345,
    "name": "John Doe",
    "email": "john.doe@example.com",
    "is_active": true,
    "age": 30
}

et un objet B :

{
    "id": 12345,
    "name": "James Watt",
    "email": "james.watt@example.com",
    "is_active": false,
    "age": 34
}

On veut supprimer les objets ayant le même id.

In [9]:
#tag::delete_duplicate_according_one_key[]
def delete_duplicate_according_one_key(data, key):
    """Delete duplicate according one key"""

    unique_values = set()
    output_data = []

    for entry in data:
        key_value = entry.get(key)  # Replace 'key' with the actual key name
        if key_value not in unique_values:
            unique_values.add(key_value)
            output_data.append(entry)
#end::delete_duplicate_according_one_key[]

## Supprimer des doublons parfaits

In [10]:
#tag::delete_duplicate_according_all_keys[]
def delete_duplicate_according_all_keys(input_file_path, output_file_path):
    """Delete duplicate according all keys"""

    with open(input_file_path, 'r') as f:
        data = json.load(f)

    # Remove duplicate
    data = [i for n, i in enumerate(data) if i not in data[n + 1:]]

    with open(output_file_path, 'w') as f:
        json.dump(data, f, indent=4, sort_keys=True)